zcash_client_sqlite/src/wallet.rs

```diff
@@ -491,56 +493,266 @@ pub(crate) fn is_valid_account_extfvk<P: consensus::Parameters>(
+            Ok(conn
+                .query_row(
+                    "SELECT MIN(shard_index), MAX(shard_index)
+                     FROM sapling_tree_shards
+                     WHERE subtree_end_height > :fully_scanned_height
+                     OR subtree_end_height IS NULL",
+                    named_params![":fully_scanned_height": u32::from(fully_scanned_height)],
+                    |row| {
+                        let min_tree_size = row
+                            .get::<_, Option<u64>>(0)?
+                            .map(|min| min << SAPLING_SHARD_HEIGHT);
+                        let max_idx = row.get::<_, Option<u64>>(1)?;
+                        Ok(fully_scanned_size.or(min_tree_size).zip(max_idx).map(
+                            |(min_tree_size, max)| {
+                                let max_tree_size = (max + 1) << SAPLING_SHARD_HEIGHT;
+                                Ratio::new(
+                                    scanned_count.unwrap_or(0),
+                                    max_tree_size - min_tree_size,
+                                )
+                            },
+                        ))
+                    },
+                )
+                .optional()?
+                .flatten())
+        }
     }
 }
 
-/// Returns the verified balance for the account at the specified height,
-/// This may be used to obtain a balance that ignores notes that have been
-/// received so recently that they are not yet deemed spendable.
-pub(crate) fn get_balance_at(
+/// Returns the spendable balance for the account at the specified height.
+///
+/// This may be used to obtain a balance that ignores notes that have been detected so recently
+/// that they are not yet spendable, or for which it is not yet possible to construct witnesses.
+pub(crate) fn get_wallet_summary(
     conn: &rusqlite::Connection,
-    account: AccountId,
-    anchor_height: BlockHeight,
-) -> Result<Amount, SqliteClientError> {
-    let balance = conn.query_row(
-        "SELECT SUM(value) FROM sapling_received_notes
-        INNER JOIN transactions ON transactions.id_tx = sapling_received_notes.tx
-        WHERE account = ? AND spent IS NULL AND transactions.block <= ?",
-        [u32::from(account), u32::from(anchor_height)],
-        |row| row.get(0).or(Ok(0)),
+    min_confirmations: u32,
+    progress: &impl ScanProgress,
+) -> Result<Option<WalletSummary>, SqliteClientError> {
+    let chain_tip_height = match scan_queue_extrema(conn)? {
+        Some((_, max)) => max,
+        None => {
+            return Ok(None);
+        }
+    };
+
+    let birthday_height =
+        wallet_birthday(conn)?.expect("If a scan range exists, we know the wallet birthday.");
+
+    let fully_scanned_height =
+        block_fully_scanned(conn)?.map_or(birthday_height - 1, |m| m.block_height());
+    let summary_height = (chain_tip_height + 1).saturating_sub(std::cmp::max(min_confirmations, 1));
```

Comment:
There is a slight issue here regarding notes that are just about to expire. Without the `max(_, 1)`, with `min_confirmations == 0` this would set `summary_height` to the "mempool height" (`chain_height + 1`), which is precisely what we want for any transaction in the mempool (that still has a chance of being mined, and thus mined before expiry). Down below in the note selection code, we exclude transactions that have expired as of `summary_height` (because `expiry_height < summary_height`) from the balance, which means if someone calls this API with `min_confirmations == 0`, then about-to-expire funds will disappear from the wallet balance, only to reappear again if the transaction does actually get mined in the last block it can.

I'm fine with this being an edge case we don't handle yet, and instead merging @daira's suggested comment change that we treat `min_confirmations == 0` as `min_confirmations == 1` for shielded notes.

Reply:
Filed as #940.